lib/iris/analysis/stats.py

```diff
@@ -84,18 +102,48 @@ def pearsonr(cube_a, cube_b, corr_coords=None, weights=None, mdtol=1.):
     if corr_coords is None:
         corr_coords = common_dim_coords
 
-    # Broadcast weights to shape of cube_1 if necessary.
-    if weights is None or cube_1.shape == cube_2.shape:
+    smaller_shape = cube_2.shape
+
+    # Match up data masks if required.
+    if common_mask:
+        # Create a cube of 1's with a common mask.
+        if ma.is_masked(cube_2.data):
+            mask_cube = _ones_like(cube_2)
+        else:
+            mask_cube = 1.
+        if ma.is_masked(cube_1.data):
+            # Take a slice to avoid unnecessary broadcasting of cube_2.
+            slice_coords = [dim_coords_1[i] for i in range(cube_1.ndim) if
+                            dim_coords_1[i] not in common_dim_coords and
+                            np.array_equal(cube_1.data.mask.any(axis=i),
+                                           cube_1.data.mask.all(axis=i))]
+            cube_1_slice = cube_1.slices_over(slice_coords).next()
```

Comment:
Writing this as `cube_1_slice = next(cube_1.slices_over(slice_coords))` would allow this to be Python 3 compatible.